Nodejs/Product/TestAdapter/TestFilesUpdateWatcher.cs

```diff
@@ -43,84 +39,71 @@ public bool AddDirectoryWatch(string path)
             }
             return false;
         }
 
-        public void RemoveWatch(string path)
+        public bool RemoveFileWatch(string path)
         {
             ValidateArg.NotNull(path, "path");
 
-            if (!string.IsNullOrEmpty(path))
+            if (this.watchedFiles.TryGetValue(path, out uint cookie))
             {
-                FileSystemWatcher watcher;
-
-                if (_fileWatchers.TryGetValue(path, out watcher))
-                {
-                    watcher.EnableRaisingEvents = false;
-
-                    _fileWatchers.Remove(path);
-
-                    watcher.Changed -= OnChanged;
-                    watcher.Dispose();
-                }
+                this.watchedFiles.Remove(path);
+                return ErrorHandler.Succeeded(this.fileWatcher.UnadviseFileChange(cookie));
             }
+            return false;
         }
 
-        private void OnChanged(object sender, FileSystemEventArgs e)
+        public bool RemoveFolderWatch(string path)
         {
-            var evt = FileChangedEvent;
-            if (evt != null)
+            ValidateArg.NotNull(path, "path");
+
+            if (this.watchedFiles.TryGetValue(path, out uint cookie))
             {
-                evt(sender, new TestFileChangedEventArgs(null, e.FullPath, TestFileChangedReason.Changed));
+                this.watchedFiles.Remove(path);
+                return ErrorHandler.Succeeded(this.fileWatcher.UnadviseDirChange(cookie));
             }
+            return false;
         }
 
-        private void OnRenamed(object sender, RenamedEventArgs e)
+        int IVsFileChangeEvents.FilesChanged(uint cChanges, string[] rgpszFile, uint[] rggrfChange)
         {
-            var evt = FileChangedEvent;
-            if (evt != null)
+            for (var i = 0; i < cChanges; i++)
             {
-                evt(sender, new TestFileChangedEventArgs(null, e.FullPath, TestFileChangedReason.Renamed));
+                FileChangedEvent?.Invoke(this, new TestFileChangedEventArgs(null, rgpszFile[i], ConvertVSFILECHANGEFLAGS(rggrfChange[i])));
             }
+
+            return VSConstants.S_OK;
         }
 
-        public void Dispose()
+        int IVsFileChangeEvents.DirectoryChanged(string pszDirectory)
         {
-            Dispose(true);
-            // Use SupressFinalize in case a subclass
-            // of this type implements a finalizer.
-            GC.SuppressFinalize(this);
+            FileChangedEvent?.Invoke(this, new TestFileChangedEventArgs(null, pszDirectory, TestFileChangedReason.Changed));
+            return VSConstants.S_OK;
         }
 
-        protected virtual void Dispose(bool disposing)
+        private static TestFileChangedReason ConvertVSFILECHANGEFLAGS(uint flag)
         {
-            if (disposing && _fileWatchers != null)
+            if ((flag & (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Attr)) != 0)
```

Comment:
Alternatively, eliminate this case and put the condition in the assert before `return TestFileChangedReason.Changed`.